### ISSUE TITLE
【Bug】torch.compile not support weight norm

### ISSUE THREAD

**Issue by fafancier** [pytorch/pytorch#118742]
### 🐛 Describe the bug

I found torch.comiple not support weight_norm. is there any plan to fix it ? 
Thanks in advance.

```
torch                        2.2.0
torchaudio                   2.2.0
torchvision                  0.17.0
```

### test code 

``` python
import torch
import torch.nn as nn
print("torch version: ",torch.__version__)

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.lin  = nn.Linear(10, 30)
        self.lin  = nn.utils.weight_norm(self.lin,dim=0)
    def forward(self, inp1):
        print("inp1 size :", inp1.size())
        out = self.lin(inp1)
        return out
inputs = [ torch.randn(10) ]
model = Model().to(torch.device("cpu"))
print('==== Eager mode ====')
ret_eager = model(*inputs)

print('==== TorchComp mode ====')
ret_exported = torch.compile(model)(*inputs)
print('OK!')
```

### output 
```shell

torch version:  2.2.0
/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/nn/utils/weight_norm.py:28: UserWarning: torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.
  warnings.warn("torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.")
==== Eager mode ====
inp1 size : torch.Size([10])
==== TorchComp mode ====
inp1 size : torch.Size([10])
[2024-01-31 21:24:19,961] [1/0_1] torch._subclasses.fake_tensor: [ERROR] fake tensor raised TypeError
[2024-01-31 21:24:19,961] [1/0_1] torch._subclasses.fake_tensor: [ERROR] Traceback (most recent call last):
[2024-01-31 21:24:19,961] [1/0_1] torch._subclasses.fake_tensor: [ERROR]   File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py", line 1392, in __torch_dispatch__
[2024-01-31 21:24:19,961] [1/0_1] torch._subclasses.fake_tensor: [ERROR]     return self.dispatch(func, types, args, kwargs)
[2024-01-31 21:24:19,961] [1/0_1] torch._subclasses.fake_tensor: [ERROR]   File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py", line 1604, in dispatch
[2024-01-31 21:24:19,961] [1/0_1] torch._subclasses.fake_tensor: [ERROR]     return decomposition_table[func](*args, **kwargs)
[2024-01-31 21:24:19,961] [1/0_1] torch._subclasses.fake_tensor: [ERROR] TypeError: _weight_norm_interface() missing 1 required positional argument: 'dim'
Traceback (most recent call last):
  File "/home/xxxxxx/project/ddp_neus/test.py", line 20, in <module>
    ret_exported = torch.compile(model)(*inputs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 489, in _fn
    return fn(*args, **kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xxxxxx/project/ddp_neus/test.py", line 11, in forward
    print("inp1 size :", inp1.size())
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 655, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 727, in _convert_frame
    result = inner_convert(frame, cache_entry, hooks, frame_state)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 383, in _convert_frame_assert
    compiled_product = _compile(
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 646, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 244, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 562, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py", line 1033, in transform_code_object
    transformations(instructions, code_options)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 151, in _fn
    return fn(*args, **kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 527, in transform
    tracer.run()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2128, in run
    super().run()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 818, in run
    and self.step()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 781, in step
    getattr(self, inst.opname)(inst)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 470, in wrapper
    return inner_fn(self, inst)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1213, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 652, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/nn_module.py", line 702, in call_function
    return variables.UserFunctionVariable(fn, source=source).call_function(
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 248, in call_function
    return super().call_function(tx, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 81, in call_function
    return tx.inline_user_function_return(
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 688, in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2261, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2376, in inline_call_
    tracer.run()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 818, in run
    and self.step()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 781, in step
    getattr(self, inst.opname)(inst)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 470, in wrapper
    return inner_fn(self, inst)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1213, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 652, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/user_defined.py", line 435, in call_function
    return self.call_method(tx, "__call__", args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/user_defined.py", line 321, in call_method
    return UserMethodVariable(method, self, source=source).call_function(
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 294, in call_function
    return super().call_function(tx, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 248, in call_function
    return super().call_function(tx, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 81, in call_function
    return tx.inline_user_function_return(
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 688, in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2261, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2376, in inline_call_
    tracer.run()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 818, in run
    and self.step()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 781, in step
    getattr(self, inst.opname)(inst)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 470, in wrapper
    return inner_fn(self, inst)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1213, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 652, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 294, in call_function
    return super().call_function(tx, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 248, in call_function
    return super().call_function(tx, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 81, in call_function
    return tx.inline_user_function_return(
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 688, in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2261, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2376, in inline_call_
    tracer.run()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 818, in run
    and self.step()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 781, in step
    getattr(self, inst.opname)(inst)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 470, in wrapper
    return inner_fn(self, inst)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1213, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 652, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/torch.py", line 542, in call_function
    tensor_variable = wrap_fx_proxy(
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/builder.py", line 1314, in wrap_fx_proxy
    return wrap_fx_proxy_cls(target_cls=TensorVariable, **kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/variables/builder.py", line 1399, in wrap_fx_proxy_cls
    example_value = get_fake_value(proxy.node, tx, allow_non_graph_fake=True)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 1525, in get_fake_value
    raise TorchRuntimeError(str(e)).with_traceback(e.__traceback__) from None
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 1486, in get_fake_value
    ret_val = wrap_fake_exception(
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 1027, in wrap_fake_exception
    return fn()
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 1487, in <lambda>
    lambda: run_node(tx.output, node, args, kwargs, nnmodule)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 1592, in run_node
    raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 1571, in run_node
    return node.target(*args, **kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py", line 1392, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py", line 1604, in dispatch
    return decomposition_table[func](*args, **kwargs)
torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in method _weight_norm of type object at 0x7f89c5ad8840>(*(Parameter(FakeTensor(..., size=(30, 10), requires_grad=True)), Parameter(FakeTensor(..., size=(30, 1), requires_grad=True)), 0), **{}):
_weight_norm_interface() missing 1 required positional argument: 'dim'

from user code:
   File "/home/xxxxxx/project/ddp_neus/test.py", line 12, in resume_in_forward
    out = self.lin(inp1)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1550, in _call_impl
    args_result = hook(self, args)
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/nn/utils/weight_norm.py", line 65, in __call__
    setattr(module, self.name, self.compute_weight(module))
  File "/home/xxxxxx/miniconda3/envs/tc2/lib/python3.10/site-packages/torch/nn/utils/weight_norm.py", line 24, in compute_weight
    return _weight_norm(v, g, self.dim)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```

### Versions

```
Collecting environment information...
PyTorch version: 2.2.0
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 10.5.0-1ubuntu1~22.04) 10.5.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-15-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090
GPU 4: NVIDIA GeForce RTX 3090
GPU 5: NVIDIA GeForce RTX 3090
GPU 6: NVIDIA GeForce RTX 3090

Nvidia driver version: 535.146.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True


Versions of relevant libraries:
[pip3] facenet-pytorch==2.5.3
[pip3] numpy==1.26.3
[pip3] onnxruntime-gpu==1.16.3
[pip3] torch==2.2.0
[pip3] torchaudio==2.2.0
[pip3] torchvision==0.17.0
[pip3] triton==2.2.0
[conda] blas                      1.0                         mkl  
[conda] facenet-pytorch           2.5.3                    pypi_0    pypi
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.8           py310h5eee18b_0  
[conda] mkl_random                1.2.4           py310hdb19cb5_0  
[conda] numpy                     1.26.3          py310h5f9d8c6_0  
[conda] numpy-base                1.26.3          py310hb5e798b_0  
[conda] pytorch                   2.2.0           py3.10_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.2.0               py310_cu118    pytorch
[conda] torchtriton               2.2.0                     py310    pytorch
[conda] torchvision               0.17.0              py310_cu118    pytorch

```

cc @ezyang @msaroufim @albanD @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

**Comment by lezcano** [comment]
That function has been deprecated for quite some time. Now, this also fails with `torch.nn.utils.parametrizations.weight_norm` with the same error in master, so the issue is real.

**Comment by fafancier** [comment]
Thank&nbsp;you&nbsp;for&nbsp;your&nbsp;reply.&nbsp;&nbsp;


---Original---
From: &quot;Mario Lezcano ***@***.***&gt;
Date: Wed, Jan 31, 2024 21:53 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [pytorch/pytorch] 【Bug】torch.compile not support weight norm (Issue #118742)




 
That function has been deprecated for quite some time. Now, this also fails with torch.nn.utils.parametrizations.weight_norm with the same error in master, so the issue is real.
 
&mdash;

**Comment by gau-nernst** [comment]
I just found out this regression compared to v2.1. Thanks for the quick fix. Will this fix make it to v2.2.1 release?
